Report configuration settings in the outcome file

gilles-peskine-arm opened the pull request:
Record compile-time configuration options (boolean options only) for a test run in the outcome file. This gives us a handy way to know what configuration settings each configuration name corresponds to in the outcome file.

Accomplishes https://github.com/Mbed-TLS/mbedtls/issues/3417 since `analyze_outcomes` will now warn (soon err) about options that are not getting enabled in any tests. But before marking that as resolved, we should remove test cases that we don't intend to pass, of which I'm sure there are a few, and we should add test jobs for missing coverage. I intend to do that in one or more follow-up pull requests.

This pull request adds a couple of test cases for combinations of options. Listing all interesting combinations of options is out of scope here, I think that should happen organically when we realize that a combination is of immediate interest.

## PR checklist

- [x] **changelog** no (test only)
- [x] **3.6 backport** https://github.com/Mbed-TLS/mbedtls/pull/9301
- [x] **2.28 backport** https://github.com/Mbed-TLS/mbedtls/pull/9316
- [x] Framework: https://github.com/Mbed-TLS/mbedtls-framework/pull/28
- [x] **tests** provided

ronald-cron-arm reviewed (changes requested):
This looks quite good to me. I have a few comments, questions.

gilles-peskine-arm commented on tests/CMakeLists.txt:
```diff
@@ -30,6 +30,18 @@ execute_process(
 string(REGEX REPLACE "[^;]*/" ""
        base_bignum_generated_data_files "${base_bignum_generated_data_files}")
 
+execute_process(
```
https://github.com/Mbed-TLS/mbedtls/issues/9194

gilles-peskine-arm commented on tests/Makefile:
```diff
@@ -59,6 +66,8 @@ generated_bignum_test_data: ../scripts/mbedtls_dev/test_data_generation.py
 generated_bignum_test_data:
 	echo "  Gen   $(GENERATED_BIGNUM_DATA_FILES)"
 	$(PYTHON) scripts/generate_bignum_tests.py
+.SECONDARY: generated_bignum_test_data
```
5df77c63fba92d58c76161c959aa1db337db57c6

https://github.com/Mbed-TLS/mbedtls/issues/9193

gilles-peskine-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,157 @@
+    'MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS': 'MBEDTLS_PSA_CRYPTO_C',
+}
+
+def find_super_option(cfg: config.Config,
```
It may return several options, but that's an uncommon case, and the return value is a string, not a list. So I prefer the singular. But I admit the naming isn't very good and more generally the terminology is somewhat inconsistent in this file, between “option”, “setting”, “dependency”… I'm open to more general suggestions to improve the terminology.

ronald-cron-arm replied:
I've looked at this a bit more and I've proposed several changes from "option" to "setting". But I am approving the PR anyway. If you find those suggestions useful, great otherwise never mind.

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,96 @@
+from mbedtls_dev import test_data_generation
+
+
+def single_option_case(setting: config.Setting, when_on: bool,
```
Here we are using "setting" but in the function name thus I would rather go for "single_setting_case" for the function name. That would keep things more aligned with the naming chosen in config.py (config.Setting) I'd say.

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,165 @@
+        return m.group('prefix') + 'BASIC'
+    return None
+
+def conditions_for_option(cfg: config.Config,
```
```suggestion
def conditions_for_setting(cfg: config.Config,
```

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,165 @@
+    yield [], ''
+
+
+def enumerate_boolean_option_cases(cfg: config.Config
```
```suggestion
def enumerate_single_setting_cases(cfg: config.Config
```
to align with `single_setting_case`.

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,165 @@
+    'MBEDTLS_PSA_ASSUME_EXCLUSIVE_BUFFERS': 'MBEDTLS_PSA_CRYPTO_C',
+}
+
+def find_super_option(cfg: config.Config,
```
```suggestion
def find_super_setting(cfg: config.Config,
```

ronald-cron-arm reviewed (approved):
LGTM, thanks.

gilles-peskine-arm commented:
Thanks for the suggestions. I've gone and fully changed “option” to “setting”, and moved to “dependency” for a list of optionally-negated settings.

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -105,60 +111,60 @@ def find_super_option(cfg: config.Config,
         # We don't test alt implementations, except (most) platform alts
         return
-    super_setting = find_super_option(cfg, setting)
+    super_setting = dependencies_of_setting(cfg, setting)
```
```suggestion
    dependencies = dependencies_of_setting(cfg, setting)
```

gilles-peskine-arm replied:
I amended the last commit to take care of this.

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,171 @@
+    together with a human-readable explanation of each dependency. Some
+    typical cases:
+
+    * By default, generate a one-element stream with no extra dependencies.
```
```suggestion
    * By default, generate a one-element stream with no extra dependency.
```
?

ronald-cron-arm commented on tests/scripts/generate_config_tests.py:
```diff
@@ -0,0 +1,171 @@
+    * If the setting is ignored unless some other setting is enabled, generate
+      a one-element stream with that other setting as an extra dependency.
+    * If the setting is known to interact with some other setting, generate
+      a stream with one element where this setting is on and one where it's off.
```
```suggestion
      a stream with one element where this setting is on and one where it's off.
```
This is not done here or not done here yet right? If not done yet probably better to mention it.

gilles-peskine-arm replied:
There are no cases yet where `conditions_for_setting` returns more than one condition, but the function can do so and its caller supports it. So it is part of the documentation, as a specification of what callers must expect.

ronald-cron-arm reviewed (changes requested):
A few final comments following the rewording.

gilles-peskine-arm commented:
I've rebased on top of the move of `generate_test_*` to the framework, so now there is a companion framework PR https://github.com/Mbed-TLS/mbedtls-framework/pull/28 for the new file `generate_config_tests.py` . I've only done some very basic local testing, the CI might catch more stuff.

ronald-cron-arm reviewed (approved):
I've checked the rebase and the move of generate_config_tests.py to the framework. LGTM.

paul-elliott-arm reviewed (approved):
LGTM

gilles-peskine-arm commented:
I have rebased on top of the latest split-related work, and updated the path to `psa/crypto_config.h` in the new script. This ready for re-review as well as the 3.6 and framework PR. The 2.28 backport will come shortly.

ronald-cron-arm reviewed (approved):
I've checked the rebase against the merge of #9251, LGTM.

paul-elliott-arm reviewed (approved):
LGTM

ronald-cron-arm commented:
Ah sorry, there is `framework` conflict now. You probably need to rebase your mbedtls-framework PR.

gilles-peskine-arm commented:
The framework PR is merged. The problem is in the parent repository, where `development` says the framework is moving from A to B and this PR says the framework is moving from A to C. The fact that B is an ancestor of C in the framework doesn't enter into consideration. I need to either rebase or merge this PR so that the git knows which framework version to end with.

gilles-peskine-arm commented:
> The fact that B is an ancestor of C in the framework doesn't enter into consideration.

Actually, the command line git tool is smart enough to merge automatically in this case. But the git implementation on GitHub isn't. I just observed this experimentally and [this blog post explains it](https://lucasoshiro.github.io/posts-en/2022-03-12-merge-submodule/#resolving-submodule-conflicts).

So I've pushed a merge commit with `development`.

ronald-cron-arm reviewed (approved):
LGTM
